Introduce ukarch_random and add implementation for arm64 (FEAT_RNG)

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://docs.unikraft.org/contribute.html

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://docs.unikraft.org/contribute.html) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [`arm64`, `x86_64`]
 - Platform(s): [N/A]
 - Application(s): [N/A]

### Additional configuration

The ukarch_random API is unconditionally present. Each architecture's support is conditional and can be checked via the `ukararch_random_avail()`.

In arm64, availability is controlled by a newly introduced parameter `CONFIG_ARM_64_FEAT_RNG`.

### Description of changes

#### The ukarch_random API

Processor-supported random number generation can be useful for different purposes on the platform and library level. Some examples include key / seed generation for hardware-based security protections, such as PAuth, MTE in Armv8, or seed generation for the ChaCha20-based PRNG in the ukswrand library.

This PR introduces a minimal API for processor-supported randomness, namely `ukarch_random`. The API consists of merely three functions, that map closely to the functionality provided by the relevant mechanisms in Armv8 and x86_64: `ukarch_random_avail()` checks whether the feature is implemented by the processor, `ukarch_random_get_long()` returns a random long integer, and `ukarch_random_long_reseed()` returns a random long integer after reseeding the RNG.

Given that the trustworthiness of processor-provided randomness is a controversial topic, higher layers shall ideally use `ukarch_random` optionally and when possible provide the user with additional options for other sources of randomness, such as drivers to TRNGs or Crypto Engines external to the processor.

#### Implementation in Armv8

Processor-supported randomness (FEAT_RNG) is introduced as an OPTIONAL extension in Armv8.5-a. The architecture defines the RNG as the output of a Deterministic Random Bit Generator (DRBG) that produces random numbers from a cryptographically secure algorithm, and is seeded by a True Random Number Generator (TRNG). The architecture requires that the DRBG and TRNG conform to a set of standards. Specifically the DRBG should conform to NIST SP800-90A Rev1, while TRNG should conform to NIST SP800-90B, NIST SP800-22, FIPS 140-2, BSI AIS-31.

The RNG is exposed to the user through the RNDR and RNDRSS registers. The RNDR outputs random numbers, and is re-seeded at an IMPLEMENTATION DEFINED rate. RNDRSS re-seeds the DRBG immediately before generating a new random number.

This PR provides an implementation of the `ukarch_random` API via the RNDR / RNDRSS registers, along with a new option `CONFIG_ARM_64_FEAT_RNG`, to enable the feature. As no Armv8 IP provided by Arm implements Armv8.5-A, this can be only tested in QEMU with `-machine virt -cpu max` or third-party SoCs that implement `FEAT_RNG`.

***Notice:*** Since the rate at which TRNGs can generate random bits is limited, RNDR and RNDRSS try to return a random number at a "reasonable period of time". When that fails, the PSTATE.NZCV bits are set to 0b0100, and a zero value is returned. Callers must therefore always check the result of the operation, and act as required on failure.

michpappas committed:
include/uk/arch: Introduce ukarch_random functions

This commit introduces a minimal set of functions for random number
generation implemented in the processor. The API provides functions
for initialization of the RNG, and the generation of random numbers
with and without reseeding the RNG.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

samoilescusebastian reviewed (commented):
There are some conflicts that should be resolved. Otherwise, the PR looks fine to me.

michpappas committed:
include/essentials: Provide __check_result macro

Provide a marco for the `warn_unused_result` function attribute of GCC.
Functions defined with this attribute will issue a compile-time warning
when called and the return value is unused.

This is useful for preventing bugs, especially in security-critical
functions.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas commented:
@samoilescusebastian thanks for looking into this. I recently put a bit more thinking into it, and just pushed an update with the following changes:
* Rebased to staging.
* Added the `__check_result` macro (see the relevant commit). This is useful for catching bugs, but we should make sure it's not abused, [to avoid similar problem linux folks had before](https://lwn.net/Articles/204682/). I propose that we only use it in security-critical functions. CC: @marcrittinghaus, @skuenzer.
* Replaced `ukarch_random_avail()` with `ukarch_random_init()`. This allows more generic initialization, and the implementation can be reduced to a simple check of availability if no further initialization is needed.
* Renamed functions to use fixed sizes instead of generic types, ie `ukarch_random_get_long()` now becomes `ukarch_random_u64()`. I also added `ukarch_random_u32()`. Macros for generic types can be provided in the platform layer (not part of this PR - we need to put a bit more thinking into it).
* Added `CONFIG_UKARCH_RANDOM` and stubs when not available. This will also be useful later at the platform layer.

samoilescusebastian commented:
Green light :)
However looks like `concourse-ci/app-helloworld-staging/arm-linuxu/build`   test has failed, but I don't have access to the details.

michpappas commented:
Rebased to `staging`

marcrittinghaus commented:
@michpappas

I had a quick look and it seems that the changes you mentioned (https://github.com/unikraft/unikraft/pull/434#issuecomment-1107448595) are not part of the current PR 👀

michpappas committed:
arch/arm64: Add processor-generated randomness (FEAT_RNG)

FEAT_RNG is introduced as an OPTIONAL feature in Armv8.5-a, and specifies
a hardware-based RNG that is implemented as a cryptographically secure
Deterministic Random Bit Generator (DRBG), seeded by a TRNG.

The RNG can be accessed by the RNDR and RNDRSS registers. RNDR reseeds the
DRBG at an IMPLEMENTATION DEFINED rate. RNDRRS register that seeds the
DRBG before generating the next random number.

This commit adds a new Kconfig option to enable FEAT_RNG and implements
the ukarch_random API using RNDR / RNDRRS.

Notice that as the entropy generated by a TRNG is limited, the caller
must always check the returned value to ensure that a random number was
successfully generated.

Signed-off-by: Michalis Pappas <mpappas@fastmail.fm>

michpappas commented:
@marcrittinghaus who knows what I did back then, sorry 😅 Updated with correct version.

marcrittinghaus commented:
@michpappas Thanks for updating. Now it has a conflict due to the paging code merged 😆

michpappas commented:
Rebased to staging

marcrittinghaus commented on arch/arm/arm64/Config.uk:
```diff
@@ -161,6 +161,15 @@ menu "Armv8-A Extensions"
 	help
 		BTI protects against JOP-like attacks by placing and
 		verifying landing pads on branch targets.
+
+	config ARM64_FEAT_RNG
+	bool "Armv8.5 Random Number Generator"
+	select UKARCH_RANDOM
+	help
+		This enables the hardware RNG functionality provided by the
+		processor. The architecture specifies that this is implemented
+		by a Deterministic Random Bit Generator (DRBG) that is seeded
+		from a TRNG.
```
Same comments as to #458 apply regarding the indentation.

michpappas replied:
Ok

marcrittinghaus commented on arch/arm/arm64/include/uk/asm/random.h:
```diff
@@ -0,0 +1,91 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2022, Michalis Pappas <mpappas@fastmail.fm>.
+ *                     All rights reserved.
```
Same optional note as for #458 regarding `All rights reserved.`. Your choice 😃

michpappas replied:
Will remove, thanks

marcrittinghaus commented on include/uk/arch/random.h:
```diff
@@ -0,0 +1,113 @@
+#ifndef __UKARCH_RANDOM_H__
+#define __UKARCH_RANDOM_H__
+
+#ifndef __ASSEMBLY__
+#include <uk/arch/types.h>
+#endif /* !__ASSEMBLY__ */
```
Makes generally sense, however, the remainder of the header is not ASM compatible. Since there are no definitions that could be used in ASM anyways, it is fine to just remove the check.
```suggestion
#include <uk/arch/types.h>
```

michpappas replied:
ACK

marcrittinghaus commented on arch/Config.uk:
```diff
@@ -33,3 +33,8 @@ config STACK_SIZE_PAGE_ORDER
 		order. Stack size is equal with 2^order * page size (e.g. 4KB).
 		Only change this if you know what you're doing.
+
+config UKARCH_RANDOM
```
To be more in line with the feature set config options in `lib/Config.uk` use `UKARCH_HAVE_RANDOM`.
```suggestion
config UKARCH_HAVE_RANDOM
```

michpappas replied:
Will change, but to me the `HAVE` part seems redundant (I doubt anyone would thing that `CONFIG_RNG` disables the feature 😄). Added an item in the coding conventions document to discuss.

marcrittinghaus replied:
You are right. It is as clear as it gets also without the `HAVE`. Just suggested to be in line with the current convention. Thanks for changing and also for adding this to the document for discussion 😃

marcrittinghaus commented on include/uk/arch/random.h:
```diff
@@ -0,0 +1,113 @@
+ *
+ * @return 0 on success, negative value on failure
+ */
+extern int ukarch_random_init(void);
```
Is the extern actually necessary?
```suggestion
int ukarch_random_init(void);
```

michpappas replied:
Nope, probably left over while I was working on it. Thanks for spotting it out.

marcrittinghaus commented on include/uk/arch/random.h:
```diff
@@ -0,0 +1,113 @@
+
+/* Get a 64-bit random integer
+ *
+ * @param [in|out] val A pointer to store the generated value
```
This is only an out variable.
```suggestion
 * @param [out] val A pointer to store the generated value
```

michpappas replied:
ACK

marcrittinghaus commented on include/uk/arch/random.h:
```diff
@@ -0,0 +1,113 @@
+
+/* Get a 32-bit random integer
+ *
+ * @param [in|out] val A pointer to store the generated value
```
```suggestion
 * @param [out] val A pointer to store the generated value
```

michpappas replied:
ACK

marcrittinghaus commented on include/uk/arch/random.h:
```diff
@@ -0,0 +1,113 @@
+
+/* Reseed the RNG and get a 64-bit random integer
+ *
+ * @param [in|out] val A pointer to store the generated value
```
Is this really supplying a new seed value via the parameter? In that case, please fix the parameter description to make that clear.

Don't know if this is just my VSCode plugin being uncooperative, but syntax highlighting is not working for me with `|`. Do you have the same issue? I thus always use `,`.

```suggestion
 * @param [in,out] val A pointer to store the generated value
```
![image](https://user-images.githubusercontent.com/8928900/199756250-42c48a15-46da-4729-a31c-4d6d40e44d6f.png)
![image](https://user-images.githubusercontent.com/8928900/199756328-eeb2f025-b78a-4ea2-be7a-5756a605194f.png)

If the reseed does not take a value as input, then just remove the `in`.

michpappas replied:
No, the seed is not provided by the caller. Reseeding happens by the hardware, so it's only an `[out]` parameter. Fixed.

marcrittinghaus commented on include/uk/arch/random.h:
```diff
@@ -0,0 +1,113 @@
+	return -1;
+}
+
+static inline int __check_result ukarch_random_u32(__u32 *val __unused)
```
Since you are using `__check_result` and `__unused` you have to include `uk/essentials`.

michpappas replied:
Thanks

marcrittinghaus commented on include/uk/essentials.h:
```diff
@@ -96,6 +96,9 @@ extern "C" {
 #define __noinline             __attribute__((noinline))
 #endif
+#ifndef __check_result
+#define __check_result	__attribute__((warn_unused_result))
```
See spacing comment from #458.

michpappas replied:
Ok

marcrittinghaus reviewed (commented):
Hi @michpappas,

thanks for submitting the PR! 😃 Looks pretty good. I left some minor feedback.

michpappas commented:
@marcrittinghaus thanks for the review, pushed a new revision with all comments addressed.

marcrittinghaus commented:
@michpappas Thanks for addressing the comments 😃 It looks good now from my perspective. I am going to test it again.

@samoilescusebastian Could you also do a second pass and then approve as reviewer if everything works for you?

craciunoiuc reviewed (approved):
All good here 🎲

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

marcrittinghaus reviewed (approved):
Yeet! 🥏

Approved-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>
